Adjust heuristics for reported quota

szewai opened the pull request:
#### 222e49e258bd16148526d2fc81ea7e98f55aa39a
<pre>
Adjust heuristics for reported quota
<a href="https://bugs.webkit.org/show_bug.cgi?id=256615">https://bugs.webkit.org/show_bug.cgi?id=256615</a>
rdar://problem/109173368

Reviewed by Youenn Fablet.

In existing implementation, StorageManager API returns a fixed value when existing usage is smaller than the fixed value
(standardReportedQuota, and default value is 10 GB). This patch makes a few improvements:
1. Calculate standardReportedQuota based on originQuotaRatio and standardVolumeCapacity, instead of directly setting
value for standardReportedQuota -- standardReportedQuota = standardVolumeCapacity * originQuotaRatio.
2. Unlike defaultStandardReportedQuota, defaultStandardVolumeCapacity is different on different platforms by default,
which will make the reported quota value more reasonable.
3. Ensure reported quota is at least twice the current usage. A problem in existing implementation is client needs to
store over 10 GB to find out the quota that is bigger than 10 GB, but if client sees the reported quota as 10 GB, they
may never store over that.
4. OriginQuotaManager::reportedQuota() uses reported usage (`OriginQuotaManager::usage()`) instead of actual disk usage
(`m_usage`). Reported usage tracks the realtime usage which includes the granted space, and we use that in
quota check for storage operation.

Test: WKWebsiteDataStoreConfiguration.StandardVolumeCapacity

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::originStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::reportedQuota):
(WebKit::OriginQuotaManager::reportedQuota const): Deleted.
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration standardVolumeCapacity]):
(-[_WKWebsiteDataStoreConfiguration setStandardVolumeCapacity:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::defaultStandardVolumeCapacity):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::standardVolumeCapacity const):
(WebKit::WebsiteDataStoreConfiguration::setStandardVolumeCapacity):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/264446@main">https://commits.webkit.org/264446@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7a4ea363ce2986881125063bed72dd16ccdf1ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7593 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8045 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9236 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7782 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7605 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7788 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10657 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7726 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/8915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9345 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/6223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6915 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14614 "23 flakes 96 failures") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7038 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10432 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7531 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6150 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6869 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1821 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11080 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7262 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -543,6 +550,7 @@ OriginStorageManager& NetworkStorageManager::originStorageManager(const WebCore:
                 quota = m_originQuotaRatio.value() * volumeCapacity.value();
                 increaseQuotaFunction = { };
             }
+            standardReportedQuota *= *m_originQuotaRatio;
```
I do not understand this change, how is it possible that a reported quota (in GB) is assigned a quota ratio (value without any unit).

szewai replied:
`standardReportedQuota` is initialized as `defaultStandardCapacity`, so we multiply it with ratio if ratio is not zero here (quota = disk size * ratio)

szewai replied:
> Can we add a test?
> If the test is too long due to writing large files (somehow a good sign for the mitigation), how about adding a test API to lower the initial standardReportedQuota?

Sure, I can try adding a test with test SPI.

youennf replied:
Oh sorry, I did not see the `*=`...

youennf commented on Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
```diff
@@ -156,9 +158,9 @@ void OriginQuotaManager::resetQuotaForTesting()
+uint64_t OriginQuotaManager::reportedQuota()
 {
-    if (m_usage && *m_usage > m_standardReportedQuota)
+    if (usage() > defaultReportingRealQuotaRatio * m_standardReportedQuota)
```
The more conservative idea was that we would increase the reported quota if usage gets higher.
To get the actual quota (and so derive the disk size), the web application would need to write more than half of the standard initially reported quota.
How about something like increasing m_standardReportedQuota (say multiply it by 2) in this if branch?

szewai replied:
Do you mean change it to: `if (usage() * 2 >= m_standardReportedQuota)`?

youennf replied:
Change it to something like:
```
if (usage() > defaultReportingRealQuotaRatio * m_standardReportedQuota)
    m_standardReportedQuota *= 2; // or maybe std::min(m_quota, 2 * m_standardReportedQuota)
return std::min(m_quota, m_standardReportedQuota);
```

youennf commented on Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
```diff
@@ -39,7 +39,7 @@ class OriginQuotaManager : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
     using NotifySpaceGrantedFunction = Function<void(uint64_t)>;
     static Ref<OriginQuotaManager> create(uint64_t quota, uint64_t standardReportedQuota, GetUsageFunction&&, IncreaseQuotaFunction&& = { }, NotifySpaceGrantedFunction&& = { });
-    uint64_t reportedQuota() const;
+    uint64_t reportedQuota();
```
I would tend to go with opposite way and make usage() const, with a const_cast for lazy init.

szewai replied:
Since we are going to keep increasing `m_standardReportedQuota`, this cannot be marked const.

youennf reviewed (commented):
Can we add a test?
If the test is too long due to writing large files (somehow a good sign for the mitigation), how about adding a test API to lower the initial standardReportedQuota?

szewai committed:
Adjust heuristics for reported quota
https://bugs.webkit.org/show_bug.cgi?id=256615
rdar://problem/109173368

Reviewed by Youenn Fablet.

In existing implementation, StorageManager API returns a fixed value when existing usage is smaller than the fixed value
(standardReportedQuota, and default value is 10 GB). This patch makes a few improvements:
1. Calculate standardReportedQuota based on originQuotaRatio and standardVolumeCapacity, instead of directly setting
value for standardReportedQuota -- standardReportedQuota = standardVolumeCapacity * originQuotaRatio.
2. Unlike defaultStandardReportedQuota, defaultStandardVolumeCapacity is different on different platforms by default,
which will make the reported quota value more reasonable.
3. Ensure reported quota is at least twice the current usage. A problem in existing implementation is client needs to
store over 10 GB to find out the quota that is bigger than 10 GB, but if client sees the reported quota as 10 GB, they
may never store over that.
4. OriginQuotaManager::reportedQuota() uses reported usage (`OriginQuotaManager::usage()`) instead of actual disk usage
(`m_usage`). Reported usage tracks the realtime usage which includes the granted space, and we use that in
quota check for storage operation.

Test: WKWebsiteDataStoreConfiguration.StandardVolumeCapacity

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::originStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::reportedQuota):
(WebKit::OriginQuotaManager::reportedQuota const): Deleted.
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration standardVolumeCapacity]):
(-[_WKWebsiteDataStoreConfiguration setStandardVolumeCapacity:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::defaultStandardVolumeCapacity):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::standardVolumeCapacity const):
(WebKit::WebsiteDataStoreConfiguration::setStandardVolumeCapacity):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/264446@main
